src/lib/firebase.ts

```diff
@@ -38,11 +39,31 @@ if (dev) {
 }
 
 export const getUserData = async (path?: string) => auth.currentUser && (await get(child(ref(database), `users/${auth.currentUser.uid}/${path}`))).val();
-export const setUserData = async (path: string, value: any) => auth.currentUser && (await set(child(ref(database), `users/${auth.currentUser.uid}/${path}`), value))
+export const setUserData = async (path: string, value: any) => auth.currentUser && (await fbSet(child(ref(database), `users/${auth.currentUser.uid}/${path}`), value))
 
 export const setWillAttempLogin = browser ? (willAttempt: boolean) => localStorage.setItem('willAttemptLogin', willAttempt ? 'yes' : 'no') : () => { };
 export const willAttemptLogin = browser ? () => localStorage.getItem('willAttemptLogin') === 'yes' : () => false;
 
+export const createReadOnlyStore = <T>(path: string) => readable<null | T>(null, (set) => {
+  onValue(ref(database, path), (snapshot) => {
+    set(snapshot.val());
+  });
+});
+
+export const createWritableStore = <T>(path: string) => writable<null | T>(null, (set, update) => {
+  let initialValueLoaded = false; // Prevents `update` from writing to firebase before we got the first real value
+  onValue(ref(database, path), (snapshot) => {
+    initialValueLoaded = true;
+    set(snapshot.val());
+  });
+  update((newValue) => {
+    if (initialValueLoaded) {
+      fbSet(ref(database, path), newValue);
+    }
+    return newValue;
+  });
+})
+
```

Comment:
I'll add jsdoc to these later I think, but the gist is that they will allow creating stores for Firebase data objects, so when they update we can run reactive code. In the case of the writable store, we can write code like this (note we need to reassign the entire object every time, not update nested variables):
```
let organization = createWritableStore<Database.Organization>(`/organization/${orgId}`);
const onAddMember = (uid) => {
  $organization = { ...$organization, private: { members: { ...$organization.private.members, [uid]: { role: member } } } };
};
```
Which will update both the store and Firebase. I have not tested this yet so it is still WIP.